boruta/script.whizzml

```diff
@@ -0,0 +1,391 @@
+          out-map)))
+
+;; shadow-ds
+;; Extends a dataset adding a shadow field for each field used in input_fields.
+;; If the number of shadow fields is less than 5, fields are repeated.
+;;
+;; Inputs
+;;  ds-id: (dataset-id) Dataset ID
+;;  f-ids: (list) List of field IDs that will be used as input and extended
+;;                with shadow fields
+;;
+;; Output: (list) List of elements [dataset-id shadow-field-ids fields]
+;;   dataset-id: (dataset-id) ID of the dataset with shadow fields
+;;   shadow-fields-ids: (list) List of IDs for the newly added shadow fields
+;;   fields: (map) New fields structure for the fields in the original dataset.
+;;                 This structure is needed to know the IDs in the
+;;                 shadow dataset for the original fields
+;;
+(define (shadow-ds ds-id fields obj-name)
+  (let (f-id-names (reduce (lambda (x y) (assoc x
+                                                y
+                                                (get-in fields [y "name"])))
+                           {}
+                           (keys fields))
+        obj-f (filter-map (lambda (x _) (= (get-in fields
+                                                   [x "name"])
+                                           obj-name))
+                          f-id-names)
+        obj-id (head (keys obj-f))
+        f-id-names (filter-map (lambda (x _) (!= x obj-id)) f-id-names)
+        f-ids (keys f-id-names)
+        input-fields (cons obj-id f-ids)
+        f-ids (repeat-fields f-ids 5)
+        f-names (map (lambda (x) (str "shadow " (get f-id-names x)))
+                     f-ids)
+        flatline-fs (map (lambda (x) [(flatline
+                                        "(weighted-random-value {{x}})")
+                                      (str "shadow " (get f-id-names x))])
+                         f-ids)
+        new-fields (lambda (x) (assoc {} "field" (head x)
+                                         "name" (head (tail x))))
+        sh-ds-id (create-and-wait-dataset {"origin_dataset"
+                                           ds-id
+                                           "input_fields"
+                                           input-fields
+                                           "new_fields"
+                                           (map new-fields flatline-fs)})
+        sh-ds (fetch sh-ds-id)
+        sh-fields (get sh-ds "fields")
+        sh-f-ids (filter (lambda (x) (member? (get-in sh-fields [x "name"])
+                                              f-names))
+                         (keys sh-fields))
+        old-fields (filter-map (lambda (x _) (not (member? (get-in sh-fields
+                                                           [x "name"])
+                                                  f-names)))
+                               sh-fields))
+    [sh-ds-id sh-f-ids old-fields]))
+
+;; repeat-fields
+;; Repeats the field ID until the number of elements is the one chosen
```

Comment:
Cyclically repeats the list of field IDs until it has the desired number of elements